crates/polars-plan/src/logical_plan/optimizer/projection_pushdown/hconcat.rs

```diff
@@ -0,0 +1,70 @@
+use super::*;
+
+#[allow(clippy::too_many_arguments)]
+pub(super) fn process_hconcat(
+    proj_pd: &mut ProjectionPushDown,
+    inputs: Vec<Node>,
+    schema: SchemaRef,
+    options: HConcatOptions,
+    acc_projections: Vec<Node>,
+    projections_seen: usize,
+    lp_arena: &mut Arena<ALogicalPlan>,
+    expr_arena: &mut Arena<AExpr>,
+) -> PolarsResult<ALogicalPlan> {
+    // When applying projection pushdown to horizontal concatenation,
+    // we apply pushdown to all of the inputs using the subset of accumulated projections relevant to each input,
+    // then rebuild the concatenated schema.
+
+    let schema = if acc_projections.is_empty() {
+        schema
+    } else {
+        let mut remaining_projections: PlHashSet<Node> = acc_projections.into_iter().collect();
+
+        for input in inputs.iter() {
+            let mut input_pushdown = Vec::new();
+
+            for proj in remaining_projections.iter() {
+                let input_schema = lp_arena.get(*input).schema(lp_arena);
+                if check_input_node(*proj, input_schema.as_ref(), expr_arena) {
+                    input_pushdown.push(*proj);
+                }
+            }
+
+            let mut input_names = PlHashSet::new();
+            for proj in &input_pushdown {
+                remaining_projections.remove(proj);
+                for name in aexpr_to_leaf_names(*proj, expr_arena) {
+                    input_names.insert(name);
+                }
+            }
+            proj_pd.pushdown_and_assign(
+                *input,
+                input_pushdown,
+                input_names,
+                projections_seen,
+                lp_arena,
+                expr_arena,
+            )?;
+        }
+
+        let schema_size = inputs
```

Comment:
I think we should create a `fn merge_schemas(schemas: &[Schema])` function. That can be shared with `concat_lf_horizontal` and here.

It would also amortize the cost of materializing the schema. For some logical plan nodes this is a full new allocation.

Reply:
Good idea thanks, I've added this although made it take `&[SchemaRef]` instead to avoid schema copies.